src/authentication/UserAuthenticator.js

```diff
@@ -102,3 +106,13 @@ export class UserAuthenticator {
     }
   }
 }
+
+// Promise will be called parallel to authenticating connection
+function createConnectionTimeoutPromise() {
+  return new Promise((resolve, reject) => {
+    const id = setTimeout(() => {
+      clearTimeout(id);
+      reject(new Error(CONNECTION_FAILURE));
+    }, CONNECTION_TIMEOUT_PERIOD);
+  });
+}
```

Comment:
All looks good @andreievg. I would recommend pulling this out into a more globally used utility that can be used instead of `fetch` everywhere. That's my setup in Tupaia App - I've got a `TupaiaApi` class that you can call 'get' or 'post' on, and it will handle the fetching and timeout.